### PR TITLE
theme assets pipeline

### PR DESCRIPTION
**Pull request by gumaerc**
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [ ] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable
  - [x] Opened issue for DevOps regarding necessary configuration changes to deployed environments

#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/938

#### What's this PR do?
In https://github.com/mitodl/ocw-studio/pull/891, we set up a "waterfall" like triggering strategy where publishes would happen in this sequence:

 - Create RC in `ocw-hugo-themes`
 - `ocw-theme-assets`build triggers from merge to `release-candidate`
 - `ocw-www` build triggers from artifacts being deployed to `ol-eng-artifacts`
 - `ocw-www` build creates a static version file upon completion
 - static version file triggers all other course builds

In theory this is a good approach, but in practice we've had a number of issues:

 - The compounding number of static version files in the bucket is making the trigger for it take a longer amount of time with each run
 - Through some analysis, the unzipping of the theme assets tar file is causing a compounding issue with CPU usage

This PR adjusts our whole approach for deploying new theme assets from start to finish:

 - First of all, the pipeline definition for the theme assets build has been moved from `ol-infrastructure` to here in `ocw-studio`.
 - The pipeline is also now templated and deployed using a new management command, `upsert_theme_assets_pipeline`.  It only creates a single pipeline based on the configured `GITHUB_WEBHOOK_BRANCH`, so that everything done on RC uses the `release-candidate` branch, and everything in production uses the `release` branch.
 - The theme assets pipeline has been updated to not zip up the artifacts.  Instead, they are deployed directly to the publish buckets for draft and live, then the `webpack.json` file is uploaded to `ol-eng-artifacts`.
 - The site pipeline definition was updated so that upload of the `webpack.json` file is what triggers the course build, and `ocw-hugo-themes` is pulled in directly from git for the templates
 
#### How should this be manually tested?
 - Set the following in your `.env` file:
```
CONCOURSE_URL=http://concourse:8080
CONCOURSE_PASSWORD=test
CONCOURSE_USERNAME=test
CONCOURSE_TEAM=main
```
 - Start up `ocw-studio` with the local Concourse profile using `docker-compose --profile concourse up`
 - Run `docker-compose run --rm web ./manage.py upsert_theme_asssets_pipeline`
 - Go to http://localhost:8080 in a browser and log into your local Concourse instance with the info above
 - View the theme assets pipeline and make sure it was properly created
 - Create a test course in the `ocw-studio` UI if you haven't already
 - Run `docker-compose run --rm web ./manage.py backpopulate_pipelines --filter <your site id here>`
 - Ensure that your site pipeline was created successfully and that the triggers look correct

If you actually try running any of the pipelines you will likely encounter errors as your local instance doesn't have AWS access to the buckets required.  Further testing will need to be done once this PR is merged to RC.

### CODE REVIEW

File: content_sync/pipelines/concourse.py
Comment (mbertrand) [inline content_sync/pipelines/concourse.py:309]: May want to consider moving these to the base pipeline to avoid nearly identical code, see this PR:  https://github.com/mitodl/ocw-studio/pull/950/files#diff-979f137beba738446b3fc46c89be033a4f04e2e285ce7e68d3028b0e1b83c11fR104-R138

In this case, `ocw-theme-assets` would be a value for the `version` parameter (which I will probably change to `pipeline_name` in my PR).
For example:
```
    def _make_pipeline_config_url(self, pipeline_name: str):
        """Make URL for fetching pipeline info"""
        return f"/api/v1/teams/{settings.CONCOURSE_TEAM}/pipelines/{pipeline_name}/config?vars={self.instance_vars}"
```

and call it for `ThemeAssetsPipeline` in `upsert_theme_assets_pipeline`. like this:
```
url_path = self._make_pipeline_config_url(self.pipeline_name)
```

where `self.pipeline_name` could be set in the `init()` function to `"ocw-theme-assets"`

File: content_sync/pipelines/concourse.py
Comment (mbertrand) [inline content_sync/pipelines/concourse.py:269]: I moved these in my PR so they could be used for all concourse pipelines: https://github.com/mitodl/ocw-studio/pull/950/files#diff-979f137beba738446b3fc46c89be033a4f04e2e285ce7e68d3028b0e1b83c11fR23

File: content_sync/pipelines/concourse.py
Comment (mbertrand) [inline content_sync/pipelines/concourse.py:356]: Same as comment above (put in the base pipeline to avoid dupe or nearly identical code)

File: content_sync/pipelines/concourse.py
Comment (mbertrand) [inline content_sync/pipelines/concourse.py:274]: Also "THEME_ASSETS_PIPELINE" (currently, though that may change if one of the prior suggestions is implemented).

**Review by mbertrand — CHANGES_REQUESTED** [review]

File: content_sync/pipelines/definitions/concourse/site-pipeline.yml
Comment (mbertrand) [inline content_sync/pipelines/definitions/concourse/site-pipeline.yml:131]: What does `cp ../webpack-json/webpack.json data` do?  Isn't `webpack.json` already in the `ocw-hugo-themes` directory?

File: app.json
Comment (gumaerc) [reply app.json]: I'll rename it to `CONTENT_SYNC_THEME_PIPELINE` for now and leave any other refactoring to you

File: content_sync/pipelines/concourse.py
Comment (gumaerc) [reply content_sync/pipelines/concourse.py:269]: The thing is, the theme pipeline doesn't require the Github settings.  I don't know if you saw, but I added a utility function called `check_mandatory_settings` to utils.py to abstract the actual check, then each pipeline can define what settings are mandatory.  I think in `concourse.py` at least, it makes sense for each pipeline to include the `CONCOURSE_` prefixed settings, then define what they will use beyond that?

Comment (mbertrand) [reply content_sync/pipelines/concourse.py:269]: ok, makes sense

File: content_sync/pipelines/definitions/concourse/site-pipeline.yml
Comment (gumaerc) [reply content_sync/pipelines/definitions/concourse/site-pipeline.yml:131]: No, it won't be.  `webpack.json` is generated at build time by an `AssetsPlugin` defined in the Webpack configuration and it's deposited in the theme's `base-theme/data` folder.  The `ocw-hugo-themes` folder in this case will also now just be the raw Github repo, cloned directly from github without any built artifacts.  The `webpack.json` file is uploaded to `ol-eng-artifacts` by the theme assets pipeline upon a completed build, and it has been added to the site pipeline definition as a versioned file.  The idea is that when a new version is detected, it will trigger site builds.  Although, once we merge your mass publish pipeline we'll want to remove `trigger: true` from it.  This line simply copies it from the resource import into the site's `data` file at build time so Hugo knows what URLs to write in.

**Review by mbertrand — APPROVED** [review]
Looks good, need a salt-ops PR to add `CONTENT_SYNC_THEME_PIPELINE` and `SEARCH_API_URL` setting values, otherwise 👍

**Comment by gumaerc** [comment]
@mbertrand Yep, I'll merge it after https://github.com/mitodl/salt-ops/pull/1496 is reviewed and merged.

**Pull request merged**
